# Introduction to SQL Using Python: Filtering Data with the WHERE Statement

Structured Query Language, more popularly known as SQL is a powerful Relational Database language. This language allows you to access and manipulate data stored in a SQL database. This tutorial will show you how to access the power of SQL using Python and how data can be filtered using the WHERE statement.

The first step is to import the necessary libraries. You will need to import sqlite3, this will allow you to connect to a SQL database and run SQL queries. You will also need to import pandas. Pandas will allow you to view the queried results in a clean, easy-to-read data frame format.

In [1]:
# Import necessary libraries

import sqlite3
import pandas as pd

For this tutorial, the Football Delphi database will be used. You can download the database here, https://www.kaggle.com/laudanum/footballdelphi. Take some time to read the description to learn more about the database and tables we will be using.

Next you will need to connect to the database and create a cursor object. Later we will call the cursors execute method to run SQL queries.

In [2]:
# Connect to database
conn = sqlite3.connect('''database.sqlite''')

# Create cursor object
cur = conn.cursor()

You should now be connected to the Football Delphi database. This is a relational database. A relational database includes tables. Each table is its own unique dataset that stores information. Each table is organized by columns and rows. This database includes four tables:
    - Unique_Teams
    - Matches
    - Teams
    - Teams_in_Matches

To run a SQL query and view it in an easy to view data frame format we will use the following code:

In [ ]:
cur.execute('''Enter SQL query here;''') # Runs SQL query
data = pd.DataFrame(cur.fetchall()) # Converts SQL query results into dataframe format
data.columns = [x[0] for x in cur.description] # Labels the columns of the dataframe
data # View SQL results dataframe

For the first query, we will preview the contents of the table, __Unique_Teams__. To do this we can run the following query:

In [3]:
# View Unique_Teams dataframe

cur.execute('''SELECT * 
               FROM Unique_Teams;''')
Unique_Teams_df = pd.DataFrame(cur.fetchall())
Unique_Teams_df.columns = [x[0] for x in cur.description]
Unique_Teams_df

TeamName  Unique_Team_ID
0         Bayern Munich               1
1              Dortmund               2
2            Leverkusen               3
3            RB Leipzig               4
4            Schalke 04               5
5            M'gladbach               6
6             Wolfsburg               7
7               FC Koln               8
8            Hoffenheim               9
9                Hertha              10
10                Mainz              11
11              Hamburg              12
12        Werder Bremen              13
13        Ein Frankfurt              14
14             Augsburg              15
15             Freiburg              16
16            Stuttgart              17
17             Hannover              18
18           Ingolstadt              19
19            Darmstadt              20
20            Paderborn              21
21             Nurnberg              22
22         Braunschweig              23
23       Greuther Furth              24
24   Fortuna Dusseldorf              25
25       Kaiserslautern              26
26             St Pauli              27
27               Bochum              28
28            Karlsruhe              29
29            Bielefeld              30
..                  ...             ...
98           Man United              99
99                  QPR             100
100      Sheffield Weds             101
101             Swindon             102
102      Crystal Palace             103
103           Leicester             104
104       Nott'm Forest             105
105              Bolton             106
106       Middlesbrough             107
107               Derby             108
108          Sunderland             109
109            Barnsley             110
110            Charlton             111
111             Watford             112
112            Bradford             113
113              Fulham             114
114          Birmingham             115
115         Middlesboro             116
116           West Brom             117
117          Portsmouth             118
118              Wolves             119
119               Wigan             120
120             Reading             121
121                Hull             122
122               Stoke             123
123             Burnley             124
124           Blackpool             125
125             Swansea             126
126             Cardiff             127
127         Bournemouth             128

[128 rows x 2 columns]

The __SELECT__ statement tells the database what columns we are trying to pull from the dataset. The asterisk, (*), tells the database that we want to select all the columns available in the table.
The __FROM__ statement tells the database that we want to select data from the table, __Unique_Teams__.
The SQL query is ended by a semicolon, (;). This tells the database that you have ended your query similar to the way a period ends a sentence.

You can practice this on your own by trying to query all the contents of the __Matches__ table and comparing your query to the answer below.

In [4]:
# View Matches dataframe

cur.execute('''SELECT * 
               FROM Matches;''')
Matches_df =pd.DataFrame(cur.fetchall())
Matches_df.columns = [x[0] for x in cur.description]
Matches_df

Match_ID Div  Season        Date            HomeTeam        AwayTeam  \
0             1  D2    2009  2010-04-04          Oberhausen  Kaiserslautern   
1             2  D2    2009  2009-11-01         Munich 1860  Kaiserslautern   
2             3  D2    2009  2009-10-04       Frankfurt FSV  Kaiserslautern   
3             4  D2    2009  2010-02-21       Frankfurt FSV       Karlsruhe   
4             5  D2    2009  2009-12-06               Ahlen       Karlsruhe   
5             6  D2    2009  2010-04-03        Union Berlin       Karlsruhe   
6             7  D2    2009  2009-08-14           Paderborn       Karlsruhe   
7             8  D2    2009  2010-03-08           Bielefeld       Karlsruhe   
8             9  D2    2009  2009-09-26      Kaiserslautern       Karlsruhe   
9            10  D2    2009  2009-11-21       Hansa Rostock       Karlsruhe   
10           11  D2    2009  2009-12-19      Greuther Furth       Karlsruhe   
11           12  D2    2009  2010-04-16             Koblenz       Karlsruhe   
12           13  D2    2009  2010-03-21          Oberhausen       Karlsruhe   
13           14  D2    2009  2009-09-14             Cottbus       Karlsruhe   
14           15  D2    2009  2010-05-02            Duisburg       Karlsruhe   
15           16  D2    2009  2009-10-30  Fortuna Dusseldorf       Karlsruhe   
16           17  D2    2009  2010-01-15              Aachen       Karlsruhe   
17           18  D2    2009  2009-10-18            Augsburg       Karlsruhe   
18           19  D2    2009  2010-02-05            St Pauli       Karlsruhe   
19           20  D2    2009  2009-08-24         Munich 1860       Karlsruhe   
20           21  D1    2009  2010-02-06              Bochum      Leverkusen   
21           22  D1    2009  2009-11-22       Bayern Munich      Leverkusen   
22           23  D1    2009  2010-05-08          M'gladbach      Leverkusen   
23           24  D1    2009  2009-08-08               Mainz      Leverkusen   
24           25  D1    2009  2009-10-17             Hamburg      Leverkusen   
25           26  D1    2009  2010-04-17           Stuttgart      Leverkusen   
26           27  D1    2009  2010-03-20            Dortmund      Leverkusen   
27           28  D1    2009  2009-10-31          Schalke 04      Leverkusen   
28           29  D1    2009  2009-08-22            Freiburg      Leverkusen   
29           30  D1    2009  2010-02-21       Werder Bremen      Leverkusen   
...         ...  ..     ...         ...                 ...             ...   
24595     46745  E0    2016  2017-05-06             Swansea         Everton   
24596     46746  E0    2016  2017-05-07             Arsenal      Man United   
24597     46747  E0    2016  2017-05-07           Liverpool     Southampton   
24598     46748  E0    2016  2017-05-08             Chelsea   Middlesbrough   
24599     46749  E0    2016  2017-05-10         Southampton         Arsenal   
24600     46750  E0    2016  2017-05-12             Everton         Watford   
24601     46751  E0    2016  2017-05-12           West Brom         Chelsea   
24602     46752  E0    2016  2017-05-13         Bournemouth         Burnley   
24603     46753  E0    2016  2017-05-13            Man City       Leicester   
24604     46754  E0    2016  2017-05-13       Middlesbrough     Southampton   
24605     46755  E0    2016  2017-05-13               Stoke         Arsenal   
24606     46756  E0    2016  2017-05-13          Sunderland         Swansea   
24607     46757  E0    2016  2017-05-14      Crystal Palace            Hull   
24608     46758  E0    2016  2017-05-14           Tottenham      Man United   
24609     46759  E0    2016  2017-05-14            West Ham       Liverpool   
24610     46760  E0    2016  2017-05-15             Chelsea         Watford   
24611     46761  E0    2016  2017-05-16             Arsenal      Sunderland   
24612     46762  E0    2016  2017-05-16            Man City       West Brom   
24613     46763  E0    2016  2017-05-17         Southampt

This is the table we will use for the rest of the tutorial. There are 9 columns in the __Matches__ table. The Kaggle page this database was downloaded from describes each column as follows:

 - Match_ID (int): unique ID per match
 - Div (str): identifies the division the match was played in (D1 = Bundesliga, D2 = Bundesliga 2, E0 = English Premier League)
 - Season (int): Season the match took place in (usually covering the period of August till May of the following year)
 - Date (str): Date of the match
 - HomeTeam (str): Name of the home team
 - AwayTeam (str): Name of the away team
 - FTHG (int) (Full Time Home Goals): Number of goals scored by the home team
 - FTAG (int) (Full Time Away Goals): Number of goals scored by the away team
 - FTR (str) (Full Time Result): 3-way result of the match (H = Home Win, D = Draw, A = Away Win)
 
Perphaps we only want to select the column, __Match_ID__. To do this we will just write __Match_ID__ in our select statement instead of the asterisk. The code below shows how this is done:

In [5]:
# View Match_ID from Matches

cur.execute('''SELECT Match_ID
               FROM Matches;''')
Matches_df =pd.DataFrame(cur.fetchall())
Matches_df.columns = [x[0] for x in cur.description]
Matches_df

Match_ID
0             1
1             2
2             3
3             4
4             5
5             6
6             7
7             8
8             9
9            10
10           11
11           12
12           13
13           14
14           15
15           16
16           17
17           18
18           19
19           20
20           21
21           22
22           23
23           24
24           25
25           26
26           27
27           28
28           29
29           30
...         ...
24595     46745
24596     46746
24597     46747
24598     46748
24599     46749
24600     46750
24601     46751
24602     46752
24603     46753
24604     46754
24605     46755
24606     46756
24607     46757
24608     46758
24609     46759
24610     46760
24611     46761
24612     46762
24613     46763
24614     46764
24615     46765
24616     46766
24617     46767
24618     46768
24619     46769
24620     46770
24621     46771
24622     46772
24623     46773
24624     46774

[24625 rows x 1 columns]

Try selecting only the column __HomeTeam__ from the __Matches__ table and compare your code to the query below:

In [6]:
# Select HomeTeam from the Matches table

cur.execute('''SELECT HomeTeam
               FROM Matches;''')
Matches_df =pd.DataFrame(cur.fetchall())
Matches_df.columns = [x[0] for x in cur.description]
Matches_df

HomeTeam
0              Oberhausen
1             Munich 1860
2           Frankfurt FSV
3           Frankfurt FSV
4                   Ahlen
5            Union Berlin
6               Paderborn
7               Bielefeld
8          Kaiserslautern
9           Hansa Rostock
10         Greuther Furth
11                Koblenz
12             Oberhausen
13                Cottbus
14               Duisburg
15     Fortuna Dusseldorf
16                 Aachen
17               Augsburg
18               St Pauli
19            Munich 1860
20                 Bochum
21          Bayern Munich
22             M'gladbach
23                  Mainz
24                Hamburg
25              Stuttgart
26               Dortmund
27             Schalke 04
28               Freiburg
29          Werder Bremen
...                   ...
24595             Swansea
24596             Arsenal
24597           Liverpool
24598             Chelsea
24599         Southampton
24600             Everton
24601           West Brom
24602         Bournemouth
24603            Man City
24604       Middlesbrough
24605               Stoke
24606          Sunderland
24607      Crystal Palace
24608           Tottenham
24609            West Ham
24610             Chelsea
24611             Arsenal
24612            Man City
24613         Southampton
24614           Leicester
24615             Arsenal
24616             Burnley
24617             Chelsea
24618                Hull
24619           Leicester
24620           Liverpool
24621          Man United
24622         Southampton
24623             Swansea
24624             Watford

[24625 rows x 1 columns]

You can select multiple columns by typing the name of each column in the select statement and separating each column name by a comma, (,). You can see the names of both the Home Teams and Away Teams in each match below:

In [7]:
# View HomeTeam and AwayTeam from Matches

cur.execute('''SELECT HomeTeam,
                      AwayTeam
               FROM Matches;''')
Matches_df =pd.DataFrame(cur.fetchall())
Matches_df.columns = [x[0] for x in cur.description]
Matches_df

HomeTeam        AwayTeam
0              Oberhausen  Kaiserslautern
1             Munich 1860  Kaiserslautern
2           Frankfurt FSV  Kaiserslautern
3           Frankfurt FSV       Karlsruhe
4                   Ahlen       Karlsruhe
5            Union Berlin       Karlsruhe
6               Paderborn       Karlsruhe
7               Bielefeld       Karlsruhe
8          Kaiserslautern       Karlsruhe
9           Hansa Rostock       Karlsruhe
10         Greuther Furth       Karlsruhe
11                Koblenz       Karlsruhe
12             Oberhausen       Karlsruhe
13                Cottbus       Karlsruhe
14               Duisburg       Karlsruhe
15     Fortuna Dusseldorf       Karlsruhe
16                 Aachen       Karlsruhe
17               Augsburg       Karlsruhe
18               St Pauli       Karlsruhe
19            Munich 1860       Karlsruhe
20                 Bochum      Leverkusen
21          Bayern Munich      Leverkusen
22             M'gladbach      Leverkusen
23                  Mainz      Leverkusen
24                Hamburg      Leverkusen
25              Stuttgart      Leverkusen
26               Dortmund      Leverkusen
27             Schalke 04      Leverkusen
28               Freiburg      Leverkusen
29          Werder Bremen      Leverkusen
...                   ...             ...
24595             Swansea         Everton
24596             Arsenal      Man United
24597           Liverpool     Southampton
24598             Chelsea   Middlesbrough
24599         Southampton         Arsenal
24600             Everton         Watford
24601           West Brom         Chelsea
24602         Bournemouth         Burnley
24603            Man City       Leicester
24604       Middlesbrough     Southampton
24605               Stoke         Arsenal
24606          Sunderland         Swansea
24607      Crystal Palace            Hull
24608           Tottenham      Man United
24609            West Ham       Liverpool
24610             Chelsea         Watford
24611             Arsenal      Sunderland
24612            Man City       West Brom
24613         Southampton      Man United
24614           Leicester       Tottenham
24615             Arsenal         Everton
24616             Burnley        West Ham
24617             Chelsea      Sunderland
24618                Hull       Tottenham
24619           Leicester     Bournemouth
24620           Liverpool   Middlesbrough
24621          Man United  Crystal Palace
24622         Southampton           Stoke
24623             Swansea       West Brom
24624             Watford        Man City

[24625 rows x 2 columns]

Practice selecting multiple columns by selecting __Match_ID__ and __Date__ from the __Matches__ table and compare your query to the code below:

In [8]:
# View Match_ID and Date from Matches

cur.execute('''SELECT Match_ID,
                      Date
               FROM Matches;''')
Matches_df =pd.DataFrame(cur.fetchall())
Matches_df.columns = [x[0] for x in cur.description]
Matches_df

Match_ID        Date
0             1  2010-04-04
1             2  2009-11-01
2             3  2009-10-04
3             4  2010-02-21
4             5  2009-12-06
5             6  2010-04-03
6             7  2009-08-14
7             8  2010-03-08
8             9  2009-09-26
9            10  2009-11-21
10           11  2009-12-19
11           12  2010-04-16
12           13  2010-03-21
13           14  2009-09-14
14           15  2010-05-02
15           16  2009-10-30
16           17  2010-01-15
17           18  2009-10-18
18           19  2010-02-05
19           20  2009-08-24
20           21  2010-02-06
21           22  2009-11-22
22           23  2010-05-08
23           24  2009-08-08
24           25  2009-10-17
25           26  2010-04-17
26           27  2010-03-20
27           28  2009-10-31
28           29  2009-08-22
29           30  2010-02-21
...         ...         ...
24595     46745  2017-05-06
24596     46746  2017-05-07
24597     46747  2017-05-07
24598     46748  2017-05-08
24599     46749  2017-05-10
24600     46750  2017-05-12
24601     46751  2017-05-12
24602     46752  2017-05-13
24603     46753  2017-05-13
24604     46754  2017-05-13
24605     46755  2017-05-13
24606     46756  2017-05-13
24607     46757  2017-05-14
24608     46758  2017-05-14
24609     46759  2017-05-14
24610     46760  2017-05-15
24611     46761  2017-05-16
24612     46762  2017-05-16
24613     46763  2017-05-17
24614     46764  2017-05-18
24615     46765  2017-05-21
24616     46766  2017-05-21
24617     46767  2017-05-21
24618     46768  2017-05-21
24619     46769  2017-05-21
24620     46770  2017-05-21
24621     46771  2017-05-21
24622     46772  2017-05-21
24623     46773  2017-05-21
24624     46774  2017-05-21

[24625 rows x 2 columns]

# The WHERE Statement

There are 24,625 rows in the Matches table. This makes it difficult to find specific information. Perhaps we only wanted information from the 2015 Season. To do this, we can add a __WHERE__ statement to our SQL query. The __WHERE__ statement allows us to select information that meets a certain condition. To query the information only pertaining to the 2015 season from the __Matches__ tables, we can use the query below:

In [ ]:
# View all columns from Matches from the 2015 season

cur.execute('''SELECT *
               FROM Matches
               WHERE Season = 2015;''')
Matches_df =pd.DataFrame(cur.fetchall())
Matches_df.columns = [x[0] for x in cur.description]
Matches_df

Now we only have 992 rows and each row is from a 2015 match. This allows us to see the data from the 2015 season much easier than scrolling through the entire table as queired before. Just as the __SELECT__ statement allows us to return specified columns from the datatable, the __WHERE__ statement allows us to return specific rows from the dataset that meet a certain condition.
To practice using the __WHERE__ statement, write a query that selects all the columns from the __Matches__ table where the Full Time Home Goals (__FTHG__) is equivalent to 5 and compare your query to the query below:

In [ ]:
# View all columns from Matches where FTHG = 5

cur.execute('''SELECT *
               FROM Matches
               WHERE FTHG = 5;''')
Matches_df =pd.DataFrame(cur.fetchall())
Matches_df.columns = [x[0] for x in cur.description]
Matches_df

So far we have only used numerical data in the __WHERE__ statement. In order to use string data in the __WHERE__ statement, we need to wrap quotation marks around the data we are using as a condition. To query all the data where Arsenal is the Home Team we can use the query below:

In [ ]:
# View all data from Matches where Arsenal is the Home Team

cur.execute('''SELECT *
               FROM Matches
               WHERE HomeTeam = 'Arsenal';''')
Matches_df =pd.DataFrame(cur.fetchall())
Matches_df.columns = [x[0] for x in cur.description]
Matches_df

To practice using string data as a condition in the __WHERE__ statement, select all the columns that are in the D2 division and compare your query to the one below:

In [ ]:
# View all data from Matches where the division is D2

cur.execute('''SELECT *
               FROM Matches
               WHERE Div = 'D2';''')
Matches_df =pd.DataFrame(cur.fetchall())
Matches_df.columns = [x[0] for x in cur.description]
Matches_df

## Comparison Operators

So far, we have only used the equal sign (=) as an operator in the where statement. In addition to the equality sign you can use other opperators including, 
    - > (greater than)
    - < (less than)
    - >= (greater than or equal to)
    - <= (less than or equal to)
    - != (not equal to)
To see all the data from the __Matches__ table where the Away Team scored 3 or more times, we can use the following query:

In [ ]:
# View all data from Matches where the Away Team scored 3 or more times

cur.execute('''SELECT *
               FROM Matches
               WHERE FTAG >= 3;''')
Matches_df =pd.DataFrame(cur.fetchall())
Matches_df.columns = [x[0] for x in cur.description]
Matches_df

To practice using condtional operators in the where statement, query all the data from the __Matches__ table, where the Home Team did not win.

In [ ]:
# View all data from Matches where the Home Team did not win

cur.execute('''SELECT *
               FROM Matches
               WHERE FTR != 'H';''')
Matches_df = pd.DataFrame(cur.fetchall())
Matches_df.columns = [x[0] for x in cur.description]
Matches_df

## AND & OR: Using Multiple Conditions in the WHERE Statement 

So far we have only put one condition in the __WHERE__ statement but it is possible to put more than one condition in the __WHERE__ statement. To do this you can add an __AND__ or __OR__ clause into your SQL query to further filter your results. When using the __AND__ clause in the __WHERE__ statement, each statement on either side of the __AND__ clause must be true for an observation to be returned. To see how this works, the query below will show all observations from the __Matches__ table where Man United was the Home Team and the final outcome of the game was a draw.

In [ ]:
# View all data from Matches where the Home Team was Man United and the outcome of the game was a draw

cur.execute('''SELECT *
               FROM Matches
               WHERE HomeTeam = 'Man United' 
                   AND FTR = 'D';''')
Matches_df = pd.DataFrame(cur.fetchall())
Matches_df.columns = [x[0] for x in cur.description]
Matches_df

When using the __OR__ clause in a __WHERE__ statement, rows will only be included in the results when either the first condition is true or the second condition is true. If both conditions are true for an observation than that observation will also be returned. To see how the __OR__ clause works, view the following query that shows all the rows from the __Matches__ table that were in either the 2012 season or the 2013 season:

In [ ]:
# View all data from Matches from the 2012 and 2013 seasons

cur.execute('''SELECT *
               FROM Matches
               WHERE Season = 2012 
                   OR Season = 2013;''')
Matches_df = pd.DataFrame(cur.fetchall())
Matches_df.columns = [x[0] for x in cur.description]
Matches_df

To practice using the __AND__ clause in the __WHERE__ statement, select all the rows from the __Matches__ table where the Home Team scored more than 3 points and the Away team won the game and compare to the query below:

In [ ]:
# View all the rows from Matches where the Home Team scored more than 3 goals and the Away Team won the game

cur.execute('''SELECT *
               FROM Matches
               WHERE FTHG > 3 
                   AND FTR = 'A';''')
Matches_df = pd.DataFrame(cur.fetchall())
Matches_df.columns = [x[0] for x in cur.description]
Matches_df

To practice using the __OR__ clause in the where statement, select the HomeTeam, AwayTeam, FTHG, FTAG, FTR from the __Matches__ table where the Home Team scored more than 7 goals or the Away Team scored more than 7 goals and compare to the query below:

In [ ]:
# Show the HomeTeam, AwayTeam, FTHG, FTAG, FTR from the Matches table 
# where the Home Team scored more than 7 goals or the Away Team scored more than 7 goals

cur.execute('''SELECT HomeTeam,
                      AwayTeam,
                      FTHG,
                      FTAG,
                      FTR
               FROM Matches
               WHERE FTHG > 7 
                   OR FTAG > 7;''')
Matches_df = pd.DataFrame(cur.fetchall())
Matches_df.columns = [x[0] for x in cur.description]
Matches_df

## BETWEEN & NOT BETWEEN: Filter Rows that Fall Into a Range

Now, we will introduce two more clauses that can be included in the __WHERE__ statement, __BETWEEN__ and __NOT BETWEEN__. With a __BETWEEN__ clause, you can filter your results based on the condition that value in the column being filtered has to be between two values in order to be returned. For example, if we wanted to select all the matches from the 2012 season to all the matches from the 2015 season, we would use the following query:

In [ ]:
# View all data from Matches between the 2012 season and the 2015 season

cur.execute('''SELECT *
               FROM Matches
               WHERE Season BETWEEN 2012 
                   AND 2015;''')
Matches_df = pd.DataFrame(cur.fetchall())
Matches_df.columns = [x[0] for x in cur.description]
Matches_df

The __NOT BETWEEN__ clause works in the oppisite way. If we were to write __NOT BETWEEN__ in place of __BETWEEN__ in the query above, it would only show the matches that were not included in the 2012-2015 season range. So the matches prior to the 2012 season would be included and the matches after the 2015 season would have been inlcuded in the returned result but none of the matches between the 2012 season and the 2015 season would have been included. To query the matches that have a _Match_ID that is not between 25 and 46,750, we could use the following query: 

In [ ]:
# Select all the columns from the Matches table that show the matches that do not have a Match_ID 
# between 25 and 46,750

cur.execute('''SELECT *
               FROM Matches
               WHERE Match_ID NOT BETWEEN 25 
                   AND 46750;''')
Matches_df = pd.DataFrame(cur.fetchall())
Matches_df.columns = [x[0] for x in cur.description]
Matches_df

To practice on your own, select FTHG, FTAG, FTR of all the matches where the number of goals scored by the Home Team is between 7 and 9 AND the number of goals scored by the Away Team is not between 0 and 3. When you have done this, compare your query to the one below: 

In [ ]:
# Select FTHG, FTAG, FTR of all the matches where the number of goals 
# scored by the Home Team is between 7 and 9 AND the number of goals 
# scored by the Away Team is not between 0 and 3.

cur.execute('''SELECT FTHG,
                      FTAG,
                      FTR
               FROM Matches
               WHERE (FTHG BETWEEN 7 AND 9)
                   AND (FTAG NOT BETWEEN 0 AND 3);''')
Matches_df = pd.DataFrame(cur.fetchall())
Matches_df.columns = [x[0] for x in cur.description]
Matches_df

## IN( ): Selecting Values from a List

Sometimes you may only want to select data that is inlcuded in a certain list. This is when the __IN__ operator comes into use. You can use the __IN__ operator in the __WHERE__ statement if you want to only return rows when the condition is included in a list. For example, the following query will return all the matches where Chelsea, Hull or Watford was the Home team:

In [ ]:
# Show all matches where the Home Team was either Chelsea, Hull or Watford

cur.execute('''SELECT *
               FROM Matches
               WHERE HomeTeam IN('Chelsea', 'Hull', 'Watford');''')
Matches_df = pd.DataFrame(cur.fetchall())
Matches_df.columns = [x[0] for x in cur.description]
Matches_df

Practice using the __IN__ operator to select the AwayTeam from the __Matches__ table when the Away team is either Liverpool, Man City or Swansea and compare your query to the one below:

In [ ]:
# Select AwayTeam from Matches when the Away Team name is Liverpool, Man City or Swansea

cur.execute('''SELECT AwayTeam
               FROM Matches
               WHERE AwayTeam IN('Liverpool', 'Man City', 'Swansea');''')
Matches_df = pd.DataFrame(cur.fetchall())
Matches_df.columns = [x[0] for x in cur.description]
Matches_df

## LIKE Operator: Filtering String Data

Next, we will introduce the __LIKE__ operator. This operator is useful when you want to filter string data. To see how it works, look at the query below. This query will return all the matches where the Home Team name begins with "A".

In [ ]:
# Show all the matches where the Home Team name begins with an "A"

cur.execute('''SELECT *
               FROM Matches
               WHERE HomeTeam LIKE 'A%';''')
Matches_df = pd.DataFrame(cur.fetchall())
Matches_df.columns = [x[0] for x in cur.description]
Matches_df

After the like statement we wrapped in quotation marks, A%. The A tells us that the statement must begin with an "A". The percent sign is used like a wild card. Anything after the A is not required to be any specific value. To see how this works view the following:

 - WHERE HomeTeam LIKE 'Be%' (Returns rows where HomeTeam begins with "Be")
 - WHERE HomeTeam LIKE '%ty' (Returns rows where HomeTeam ends with "ty")
 - WHERE HomeTeam LIKE '%a%' (Returns rows where HomeTeam includes "a")
 
 To practice using the __LIKE__ statement to return all matches where the Away Team name includes the term "City" and compare to the following code:

In [ ]:
# Show all the matches where the Away Team name includes the term "City"

cur.execute('''SELECT *
               FROM Matches
               WHERE AwayTeam LIKE '%City%';''')
Matches_df = pd.DataFrame(cur.fetchall())
Matches_df.columns = [x[0] for x in cur.description]
Matches_df

You can also use the term __NOT LIKE__. If you wanted to return all the matches where the Date did not begin with 2017 you could use the following query:

In [ ]:
# Show all the matches where the Date did not begin with '2017'

cur.execute('''SELECT *
               FROM Matches
               WHERE Date NOT LIKE '2017%';''')
Matches_df = pd.DataFrame(cur.fetchall())
Matches_df.columns = [x[0] for x in cur.description]
Matches_df

To practice on your own, select all columns from the __Matches__ table where the __Date__ does not begins with 2016 and the HomeTeam name ends with a "y" and the AwayTeam name does not includes an "e". Compare your query to the one below:

In [ ]:
# Show all the matches where the Date did not begin with '2017'

cur.execute('''SELECT *
               FROM Matches
               WHERE (Date NOT LIKE '2016%')
                   AND (HomeTeam Like '%y')
                   AND (AwayTeam NOT LIKE '%e%');''')
Matches_df = pd.DataFrame(cur.fetchall())
Matches_df.columns = [x[0] for x in cur.description]
Matches_df

## IS NULL & IS NOT NULL: Dealing with Missing Values

Sometimes, data may be missing in some rows. These values are known as __NULL__ data. You can check if any of the data is missing for a specific column using the clause __IS NULL__ in the where statement. To see if the outcome of any match is missing from the Matches table we can use the following query:

In [ ]:
# Select all columns from the Matches table where FTR is a NULL value

cur.execute('''SELECT *
               FROM Matches
               WHERE FTR IS NULL;''')
Matches_df = pd.DataFrame(cur.fetchall())
Matches_df.columns = [x[0] for x in cur.description]
Matches_df

We recieved a length mismatch error. This is because there are now observations where the outcome of a game is unknown. That is good. Practice using the __IS NULL__ clause to check whether any of the Home Team Goal Count is missing from any of the observations in the __Matches__ table and compare your query to the one below:

In [ ]:
# Select all columns from the Matches table where FTHG is a NULL value

cur.execute('''SELECT *
               FROM Matches
               WHERE FTHG IS NULL;''')
Matches_df = pd.DataFrame(cur.fetchall())
Matches_df.columns = [x[0] for x in cur.description]
Matches_df

There should be no values missing. The statement __IS NOT NULL__ works in the oppisite way. This will return all values where a column does not have a NULL value. The following query shows all the observations from the __Matches__ table where there is not a null value in the __HomeTeam__ column.

In [ ]:
# Select all columns from the Matches table where HomeTeam is not a NULL value

cur.execute('''SELECT *
               FROM Matches
               WHERE HomeTeam IS NOT NULL;''')
Matches_df = pd.DataFrame(cur.fetchall())
Matches_df.columns = [x[0] for x in cur.description]
Matches_df

Practice using the __NOT NULL__ clause by querying all the columns from the Matches table where __Date__ does not have a null value and compare to the query below:

In [ ]:
# Select all columns from the Matches table where Date is not a NULL value

cur.execute('''SELECT *
               FROM Matches
               WHERE Date IS NOT NULL;''')
Matches_df = pd.DataFrame(cur.fetchall())
Matches_df.columns = [x[0] for x in cur.description]
Matches_df

An important thing to note when using __IS NULL__ and __IS NOT NULL__, a value of zero is not equivalent to a NULL value. A NULL is a missing value and a value of 0 is not a mssing value.

In [ ]:
# View Teams dataframe

cur.execute('''SELECT * FROM Teams;''')
Teams_df = pd.DataFrame(cur.fetchall())
Teams_df.columns = [x[0] for x in cur.description]
Teams_df.head()

In [ ]:
# View Teams_in_Matches dataframe

cur.execute('''SELECT * FROM Teams_in_Matches;''')
Teams_in_Matches_df = pd.DataFrame(cur.fetchall())
Teams_in_Matches_df.columns = [x[0] for x in cur.description]
Teams_in_Matches_df.head()

In [ ]:
# View Unique_Teams dataframe

cur.execute('''SELECT * FROM Unique_Teams;''')
Unique_Teams_df = pd.DataFrame(cur.fetchall())
Unique_Teams_df.columns = [x[0] for x in cur.description]
Unique_Teams_df.head()